Add internet access page (#39)
eth-cscs/cscs-docs: docs/guides/internet-access.md

[](){#ref-guides-internet-access}

> # Internet Access on Alps

Author: Maybe this is sufficient:
```suggestion
# Internet Access
```
?

Reviewer: I prefer the "on Alps"

The [Alps network][ref-alps-hsn] is mostly configured with private IP addresses (`172.28.0.0/16`).
Login nodes have public IP addresses which means that they can directly access the internet, while a proxy server provides internet access for compute nodes.

??? info "Compute node proxy configuration"

    Compute nodes are configured with the following environment variables to use the proxy server:
    
    ```bash
    https_proxy=http://proxy.cscs.ch:8080
    http_proxy=http://proxy.cscs.ch:8080
    no_proxy=.local, .cscs.ch, localhost, 148.187.0.0/16, 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16
    HTTPS_PROXY=http://proxy.cscs.ch:8080
    HTTP_PROXY=http://proxy.cscs.ch:8080
    NO_PROXY=.local, .cscs.ch, localhost, 148.187.0.0/16, 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16
    ```

## Using SSH through the proxy server 

> While use of the proxy server is transparent for most use cases, e.g. cloning git repositories from GitHub over SSH requires additional configuration for compute nodes.

Reviewer 2: ```suggestion
While use of the proxy server is transparent for most use cases, 
others need additional configuration for compute nodes.
An example is cloning git repositories from GitHub over SSH.
```
Not sure about the phrasing starting with "e.g.".

Cloning over https works without additional configuration.
To make SSH use the proxy server, add the following to your `~/.ssh/config` file:

```bash
Match Host *,!148.187.0.0/16,!192.168.0.0/16,!172.16.0.0/12,!10.0.0.0/8 exec "hostname -I | grep -vqF 148.187."
    ProxyCommand nc -X connect -x proxy.cscs.ch:8080 %h %p
```

This configuration takes into account that login and compute nodes require a different setup.

> !!! info "Error message when cloning without the proxy set up for SSH"

Reviewer 2: ```suggestion
??? info "Error message when cloning without the proxy set up for SSH"
```
Maybe this can be collapsed? I'm assuming here that collapsed blocks do show up in a search.

    When cloning a git repository without the correct SSH configuration, cloning will time out as follows:
    ```bash
    [daint][<user>@daint-ln001 ~]$ git clone git@github.com:open-mpi/ompi.git
    Cloning into 'ompi'...
    ssh: connect to host github.com port 22: Connection timed out
    fatal: Could not read from remote repository.

    Please make sure you have the correct access rights
    and the repository exists.
    ```

## Accessing the public IP of a node

> When on a login node configured with a public IP address, you can retrieve the public IP address for example as follows:

Reviewer 2: Maybe it is worth adding an info/warning somewhere pointing out that the public IP address is essentially a shared resource, and therefore limitations based on the IP address (such as pulls from DockerHub) are shared across all users?

Author: Good idea!

Author: I added a note in https://github.com/eth-cscs/cscs-docs/pull/39/commits/942d2929e163dc72f4dae62c30d51e38b61ebb8f. I put the note at the top of the page since the public IP is shared in all cases, whether on login nodes or through the proxy server. I also cross-linked a warning in the container engine docs. What do you think?

Reviewer 2: I think that's great, thanks!

[](){#ref-guides-internet-access}
# Internet Access on Alps

The [Alps network][ref-alps-hsn] is mostly configured with private IP addresses (`172.28.0.0/16`).
Login nodes have public IP addresses which means that they can directly access the internet, while a proxy server provides internet access for compute nodes.

??? info "Compute node proxy configuration"

    Compute nodes are configured with the following environment variables to use the proxy server:
    

>     ```bash

Reviewer: To make it easier for users to cut and paste this when they need to, add `export` to the start of each line.

Author: On one hand I agree, adding the export makes it easier. On the other, do they need to cut and paste if it's configured correctly? This is really just a "FYI". Which makes me think: do we need to keep the actual values? Or just say that they're set appropriately?

Reviewer: I made this comment because I got burnt by exactly this the other day when using a container - the variables don't seem to be set properly inside the container in some cases.
I wanted to cut and paste from the original docs, and was annoyed by having to manually add `export` for each line.

You could create a table with the variables, and an expandable `example` box that has the text ready to "cut and paste"?

Author: Fair enough, that's a good reason to leave them here then and add the `export`! I added the `export` in https://github.com/eth-cscs/cscs-docs/pull/39/commits/e5370975c086e777ae159a3b39538ca672d7184a. Is it worth having the extra table as well? I'd prefer not duplicating the information, even if they're close to each other and "easy" to spot if both need to be updated.

    https_proxy=http://proxy.cscs.ch:8080
    http_proxy=http://proxy.cscs.ch:8080
    no_proxy=.local, .cscs.ch, localhost, 148.187.0.0/16, 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16
    HTTPS_PROXY=http://proxy.cscs.ch:8080
    HTTP_PROXY=http://proxy.cscs.ch:8080
    NO_PROXY=.local, .cscs.ch, localhost, 148.187.0.0/16, 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16
    ```

!!! warning "Public IPs are a shared resource"
    Be aware that public IPs, whether on login nodes or through the proxy, are essentially a shared resource.
    Many services will rate limit or block usage based on the IP address if abused.
    An example is pulling container images from Docker Hub.
    [Authenticating with Docker Hub][ref-ce-third-party-private-registries] makes their rate limit apply per user instead.

## Using SSH through the proxy server 

While use of the proxy server is transparent for most use cases, others need additional configuration for compute nodes.
An example is cloning git repositories from GitHub over SSH.
Cloning over https works without additional configuration.
To make SSH use the proxy server, add the following to your `~/.ssh/config` file:

> ```bash

Reviewer: This isn't bash code, maybe use `terminal`?

You can also give it a title

```
\```terminal "title"
Match Host *,!148.187.0.0/16,!192.168.0.0/16,!172.16.0.0/12,!10.0.0.0/8 exec "hostname -I | grep -vqF 148.187."
    ProxyCommand nc -X connect -x proxy.cscs.ch:8080 %h %p
\```
```

Author: Good catch, indeed. I removed the type altogether, since it's a plain text config file, and added the filename as title in https://github.com/eth-cscs/cscs-docs/pull/39/commits/1f9a772dbbe14a70ffa03e39c65fc2ed45f1b852.

I also found this page which has way more lexers than I would've expected: https://pygments.org/docs/lexers/#. Unfortunately none for ssh config files, but I think plain text will do just fine here.

Match Host *,!148.187.0.0/16,!192.168.0.0/16,!172.16.0.0/12,!10.0.0.0/8 exec "hostname -I | grep -vqF 148.187."
    ProxyCommand nc -X connect -x proxy.cscs.ch:8080 %h %p
```

This configuration takes into account that login and compute nodes require a different setup.

> ??? info "Error message when cloning without the proxy set up for SSH"

Reviewer: `warning` instead of `info` will make it stand out more?

Author: My thinking was that it's just "info" as something to show up in search results. The whole section is a "warning" in a sense, saying "warning: to be able to clone on compute nodes you have to set up this config". That said, I'm happy with "warning" as well, really not a big deal for me.

Reviewer: You are right that it isn't a big deal.

I thought that because the box contains an description of an error and what to do, having the orange title with an exclamation mark would make it stand out more on the page.

Author: That's fair! Made it a warning in https://github.com/eth-cscs/cscs-docs/pull/39/commits/fd29d208357a6e539cfc06e1df9591dcd0b2b6c6. Didn't mean to be difficult.
